src/test/FreeC/Pass/ExportPassTests.hs

```diff
@@ -0,0 +1,117 @@
+    it "qualifies Coq identifiers of constructor entries" $ do
+      input <- expectParseTestModule ["module A where", "data A.Bar = A.Foo"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Bar" 0 ["A.Foo"]
+        _       <- defineTestCon "A.Foo" 0 "A.Bar"
+        _       <- exportPass input
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let foo = fromJust $ lookupExportedEntry IR.ValueScope
+                                                 (Qual "A" (Ident "Foo"))
+                                                 (fromJust mOutput)
+        entryIdent foo `shouldBeQualifiedWith` "A"
+    it "qualifies Coq identifiers of type synonym entries" $ do
+      input <- expectParseTestModule
+        ["module A where", "data A.Bar = A.Bar;", "type A.Foo = A.Bar"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Bar" 0 ["A.Bar"]
+        _       <- defineTestCon "A.Bar" 0 "A.Bar"
+        _       <- defineTestTypeSyn "A.Foo" [] "A.Bar"
+        _       <- exportPass input
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let foo = fromJust $ lookupExportedEntry IR.TypeScope
+                                                 (Qual "A" (Ident "Foo"))
+                                                 (fromJust mOutput)
+        entryIdent foo `shouldBeQualifiedWith` "A"
+    it "qualifies Coq identifiers of function declaration entries" $ do
+      input <- expectParseTestModule
+        ["module A where", "data A.Bar = A.Bar;", "type A.Foo = A.Bar"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Foo" 0 ["A.Foo"]
+        _       <- defineTestCon "A.Foo" 0 "A.Foo"
+        _       <- defineTestFunc "A.mkFoo" 0 "A.Foo"
+        _       <- exportPass input
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let mkFoo = fromJust $ lookupExportedEntry IR.ValueScope
+                                                   (Qual "A" (Ident "mkFoo"))
+                                                   (fromJust mOutput)
+        entryIdent mkFoo `shouldBeQualifiedWith` "A"
+    it "does not override qualification of Coq identifiers for entries" $ do
+      input  <- expectParseTestModule ["module A where", "data A.Foo = A.Foo"]
+      input' <- expectParseTestModule ["module B where", "data B.Foo = B.Foo"]
+      shouldSucceedWith $ do
+        _       <- defineTestTypeCon "A.Foo" 0 ["A.Foo"]
+        _       <- defineTestCon "A.Foo" 0 "A.Foo"
+        _       <- defineTestTypeCon "B.Foo" 0 ["B.Foo"]
+        _       <- defineTestCon "B.Foo" 0 "B.Foo"
+        _       <- exportPass input
+        _       <- exportPass input'
+        mOutput <- inEnv $ lookupAvailableModule "A"
+        let foo = fromJust $ lookupExportedEntry IR.TypeScope
+                                                 (Qual "A" (Ident "Foo"))
+                                                 (fromJust mOutput)
+        entryIdent foo `shouldBeQualifiedWith` "A"
```

Comment:
If you looked up `B.Foo` in the module interface of `A` (without doing the `interfaceExports` check in `lookupExportedEntry`), the Coq identifier would be qualified with `A` as well. The first application of the `exportPass` qualifies all entries in the environment with `A` including the entries which are defined in `B`. Thus, the entries of `B.Foo` would have to be added after the first `exportPass`.

```haskell
        _       <- defineTestTypeCon "A.Foo" 0 ["A.Foo"]
        _       <- defineTestCon "A.Foo" 0 "A.Foo"
        _       <- exportPass input
        _       <- defineTestTypeCon "B.Foo" 0 ["B.Foo"]
        _       <- defineTestCon "B.Foo" 0 "B.Foo"
        _       <- exportPass input'
``` 

However, you don't need `data B.Foo = B.Foo` in the first place. You are actually interested in a situation like `module B where import A`. The module interface for `B` should then contain an entry for `A.Foo` whose Coq identifier is still qualified with `A`.

The test described above is most likely to fail, though, because the first `exportPass` does not change the entries in the environment itself. Only the entries in the resulting module interface are qualified, entries in the environment remain unqualified. Thus, the second `exportPass` will not "see" the qualification and qualifies everything again.

I think there are two options to solve this issue.

 1. You could create the module interface for `A` in a [`moduleEnv`](https://github.com/FreeProving/free-compiler/blob/6864d3b5a33b1b1206b355a9b3b62618605d99db/src/lib/FreeC/Monad/Converter.hs#L190) and import the interface afterwards using [`importInterface`](https://github.com/FreeProving/free-compiler/blob/6864d3b5a33b1b1206b355a9b3b62618605d99db/src/lib/FreeC/Pass/ImportPass.hs#L91).
 2. You perform only one `exportPass` and add utility methods to create environment entries with qualified Coq identifiers (or qualify existing Coq identifiers).